sunxi: add support for FriendlyARM NanoPi R1

jayanta525 opened the pull request:
Specification:
- CPU: Allwinner H3, Quad-core Cortex-A7 Up to 1.2GHz
- DDR3 RAM: 512MB/1GB
- Network:
    - 10/100/1000M Ethernet x 1，
    - 10/100M Ethernet x 1
- WiFi: 802.11b/g/n, with SMA antenna interface
- USB Host: Type-A x2
- MicroSD Slot x 1
- MicroUSB: for OTG and power input
- Debug Serial Port: 3Pin 2.54mm pitch pin-header
- Power Supply: DC 5V/2A

Installation:
- Write the image file to an SD Card with dd or any image burning tool
- Boot NanoPi from the SD Card

The following features are working and tested:
- Ethernet ports
    - 10/100/1000M Ethernet
    - 10/100M Ethernet
- LED:
    - nanopi:red:status
    - nanopi:green:wan
    - nanopi:green:lan
- KEY:
    - reset

Signed-off-by: Jayantajit Gogoi <jayanta.gogoi525@gmail.com>

adrianschmutzler commented:
This is a copy of #2987

Please don't open a new PR just for changes next time. The change should be pushed to the old PR, because otherwise we will loose all the discussion history.

jayanta525 commented:
The old fork got deleted and all commits were lost. Had to make another fork.

adrianschmutzler commented:
> The old fork got deleted and all commits were lost. Had to make another fork.

"got deleted" -> I assume _you_ deleted it ... :-)

@jayanta525 Technically, it would have been possible. (Read about git rebase and git push -f.) However, it's waste of time to discuss that now. Please take care next time.

jayanta525 commented:
Accidentally deleted the original branch the pull request was created of while trying to add support for nanopi-r2s. 

Cleanup of everything, and little knowledge about git led to another pull request. Won't happen again.

DTS and uboot patch files are ported from armbian-kernel.

Build compiles successfully and nanopi-r1 boots with the image.

Everything works as expected.

adrianschmutzler commented:
Minor formal comment:
- commit title: Add->add; remove dot at the end
- Consider adding an empty line after "Specification" and "Installation", as well as inserting a space after and before the list-bullet-points for better reading.
- "LED x 3" that's quite useless. Either name the purposes of the LEDs, or remove that line. Same for keys in commit message.

adrianschmutzler commented on package/boot/uboot-sunxi/patches/250-sun8i-h3-Nano-Pi-R1-add-device-tree.patch:
```diff
@@ -0,0 +1,115 @@
+ 	sun8i-h2-plus-bananapi-m2-zero.dtb \
+ 	sun8i-h2-plus-libretech-all-h3-cc.dtb \
+ 	sun8i-h2-plus-orangepi-r1.dtb \
++	sun8i-h3-nanopi-r1.dtb \
```
That list seems to be sorted. Consider inserting at the correct position.

jayanta525 replied:
done

adrianschmutzler commented on package/boot/uboot-sunxi/patches/250-sun8i-h3-Nano-Pi-R1-add-device-tree.patch:
```diff
@@ -0,0 +1,115 @@
++ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
++ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
++ *     OTHER DEALINGS IN THE SOFTWARE.
++ */
```
Use an SPDX license identifier instead.

jayanta525 replied:
done

adrianschmutzler commented on target/linux/sunxi/image/cortex-a7.mk:
```diff
@@ -31,6 +31,16 @@ define Device/friendlyarm_nanopi-m1-plus
 endef
 TARGET_DEVICES += friendlyarm_nanopi-m1-plus
 
+define Device/friendlyarm_nanopi-r1
```
"r1" comes after "neo"?

jayanta525 replied:
alphabetic ordering applied.

adrianschmutzler commented on target/linux/sunxi/image/cortex-a7.mk:
```diff
@@ -31,6 +31,16 @@ define Device/friendlyarm_nanopi-m1-plus
+define Device/friendlyarm_nanopi-r1
+  DEVICE_VENDOR := FriendlyARM
+  DEVICE_MODEL := NanoPi R1
+  DEVICE_PACKAGES := kmod-rtc-sunxi kmod-usb2 kmod-usb-net kmod-usb-net-rtl8152 \
```
kmod-usb-net can be removed here, it will be selected by kmod-usb-net-rtl8152

jayanta525 replied:
kmod-usb-net removed

adrianschmutzler commented on target/linux/sunxi/image/cortex-a7.mk:
```diff
@@ -31,6 +31,16 @@ define Device/friendlyarm_nanopi-m1-plus
+  DEVICE_VENDOR := FriendlyARM
+  DEVICE_MODEL := NanoPi R1
+  DEVICE_PACKAGES := kmod-rtc-sunxi kmod-usb2 kmod-usb-net kmod-usb-net-rtl8152 \
+  kmod-brcmfmac kmod-leds-gpio kmod-ledtrig-heartbeat wpad-basic \
```
Use single-tab indent for the second and third line of DEVICE_PACKAGES.

jayanta525 replied:
issue was with VS Code editor not adding tab space. Single-tab indent added.

adrianschmutzler commented on target/linux/sunxi/patches-5.4/062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1.patch:
```diff
@@ -0,0 +1,210 @@
++ *     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
++ *     FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
++ *     OTHER DEALINGS IN THE SOFTWARE.
++ */
```
Please use an SPDX identifier instead.

jayanta525 replied:
added SPDX

adrianschmutzler commented on target/linux/sunxi/patches-5.4/062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1.patch:
```diff
@@ -0,0 +1,210 @@
++#include "sun8i-h3-nanopi.dtsi"
++
++/ {
++	model = "FriendlyElec NanoPi-R1";
```
Friendly ARM?
And either remove the hyphen here, or use it consistently (however, note that the compatible will still use the hyphen ...)

jayanta525 replied:
done

adrianschmutzler commented on target/linux/sunxi/patches-5.4/062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1.patch:
```diff
@@ -0,0 +1,210 @@
++	};
++
++	leds {
++		/delete-node/ status;
```
why delete this and then set it up again as LED1?

jayanta525 replied:
corrected labels, labels are according to names printed on outer casing.

adrianschmutzler commented on target/linux/sunxi/patches-5.4/062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1.patch:
```diff
@@ -0,0 +1,210 @@
++		};
++
++		led2 {
++			label = "LED2";
```
Please use the "model:color:task" label scheme.

jayanta525 replied:
done

adrianschmutzler commented on target/linux/sunxi/patches-5.4/062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1.patch:
```diff
@@ -0,0 +1,210 @@
++		led2 {
++			label = "LED2";
++			gpios = <&pio 6 11 GPIO_ACTIVE_HIGH>;
++			default-state = "off";
```
This line can be dropped, off is default.

jayanta525 replied:
okay, got it

adrianschmutzler commented on target/linux/sunxi/patches-5.4/062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1.patch:
```diff
@@ -0,0 +1,210 @@
++			label = "LED2";
++			gpios = <&pio 6 11 GPIO_ACTIVE_HIGH>;
++			default-state = "off";
++			linux,default-trigger = "netdev";
```
This won't work?

jayanta525 replied:
cleaned

adrianschmutzler commented on target/linux/sunxi/patches-5.4/062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1.patch:
```diff
@@ -0,0 +1,210 @@
++
++	r_gpio_keys {
++		compatible = "gpio-keys";
++		input-name = "k1";
```
These two lines are already defined in dtsi.

jayanta525 replied:
lines already in dtsi removed from dts

adrianschmutzler commented on target/linux/sunxi/patches-5.4/062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1.patch:
```diff
@@ -0,0 +1,210 @@
++			label = "reset";
++			linux,code = <BTN_1>;
++			gpios = <&r_pio 0 3 GPIO_ACTIVE_LOW>;
++		};
```
Any reason for redefining that, but without using the proper code?

jayanta525 replied:
added> linux,code = <KEY_RESTART>;

adrianschmutzler commented on target/linux/sunxi/patches-5.4/062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1.patch:
```diff
@@ -0,0 +1,210 @@
++
++&usbphy {
++	status = "okay";
++};
```
That's already in DTSI as well.

jayanta525 replied:
removed

adrianschmutzler commented:
Your DTS seems to contain a lot of stuff that's in the DTSI as well (I only checked the directly included one, I assume similar cases may be found for the more remotely related files.

Please tidy that up.

adrianschmutzler commented on target/linux/sunxi/patches-5.4/062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1.patch:
```diff
@@ -0,0 +1,164 @@
++		lan {
++			label = "nanopi:green:lan";
++			gpios = <&pio 0 9 GPIO_ACTIVE_HIGH>;
++			linux,default-trigger = "netdev";
```
Still seeing netdev triggers here.

adrianschmutzler commented on target/linux/sunxi/image/cortex-a7.mk:
```diff
@@ -47,6 +47,16 @@ define Device/friendlyarm_nanopi-neo-air
+  DEVICE_MODEL := NanoPi R1
+  DEVICE_PACKAGES := kmod-rtc-sunxi kmod-usb2 kmod-usb-net-rtl8152 \
+ 	kmod-brcmfmac kmod-leds-gpio kmod-ledtrig-heartbeat wpad-basic \
+ 	brcmfmac-firmware-43430-sdio
```
There seems to be a space before the tabs here. Just one tab please for the hanging indent.

jayanta525 replied:
done

adrianschmutzler commented on target/linux/sunxi/base-files/etc/board.d/02_network:
```diff
@@ -7,19 +7,22 @@
 	;;
 *)
-	ucidef_set_interface_lan 'eth0'
+	ucidef_set_interface_lan "eth0"
```
Please rebase your changes onto recent master, so this file only shows your changes.

adrianschmutzler commented:
You'll make your life much easier if you use "git rebase" instead of merging.

Also have a look at "git rebase -i" for editing/squashing

jayanta525 commented:
i messed up using git commands.

jayanta525 commented:
resolved all files, still shows branch cannot be rebased due to conflicts.

adrianschmutzler commented:
You still have a merge commit. You PR should consist only of a single commit, without any merge commit or other additional changes. Normally, just can resolve this with "git rebase -i ...".

However, I don't know how to properly remove the merge commit first (or whether this is a problem at all), as I never use merge commits for that very reason.

jayanta525 commented:
@adrianschmutzler 
-merge commit removed
-rebased with current master

jayanta525 commented:
@hauke

jayanta525 commented:
@ynezz @adrianschmutzler 
-request for review

sbilly commented:
Hi, @jayanta525 

Does it support boot from internal mmc by default?

I've manually patched the `master` branch openwrt source code, got the firmware binary `openwrt-sunxi-cortexa7-friendlyarm_nanopi-r1-ext4-sdcard.img.gz`. And I have write it to nanopi r1's internal mmc. When I reboot the nanopi r1 device, it stopped at the u-boot prompt at last. But I can boot it by input following u-boot commands:

```
setenv fdt_high ffffffff
setenv loadkernel fatload mmc 1 $kernel_addr_r uImage
setenv loaddtb fatload mmc 1 $fdt_addr_r dtb
setenv bootargs console=ttyS0,115200 earlyprintk root=/dev/mmcblk2p2 rootwait
run loadkernel && run loaddtb && bootm $kernel_addr_r - $fdt_addr_r
```

jayanta525 commented:
Hi, @sbilly 

No, the default OpenWrt image won't boot from internal mmc by default.

Sorry, I can't test this as I don't own this board, i ported supported and created this PR on request from the users at OpenWrt forums.

You could build images with modified `boot.scr`, and use `dd` to write that image to the internal mmc with an OS booted from the sd-card or the original OpenWrt image. `sysupgrade` should also work if the `boot.scr` is modified correctly.

I remember doing the same with an OrangePi R1 while booting from it's 16MB SPI flash.

sbilly commented:
kernel print when booting.

```
[    0.002878] /cpus/cpu@0 missing clock-frequency property
[    0.002915] /cpus/cpu@1 missing clock-frequency property
[    0.002943] /cpus/cpu@2 missing clock-frequency property
[    0.002972] /cpus/cpu@3 missing clock-frequency property
```

add following to `062-ARM-dts-sun8i-add-support-for-Nano-Pi-R1-openwrt-master.patch` fixed.

```
Index: linux-5.4.48/arch/arm/boot/dts/sun8i-h3.dtsi
===================================================================
--- linux-5.4.48.orig/arch/arm/boot/dts/sun8i-h3.dtsi
+++ linux-5.4.48/arch/arm/boot/dts/sun8i-h3.dtsi
@@ -78,6 +78,7 @@
                        clock-names = "cpu";
                        operating-points-v2 = <&cpu0_opp_table>;
                        #cooling-cells = <2>;
+                       clock-frequency = <1296000000>;
                };

                cpu1: cpu@1 {
@@ -88,6 +89,7 @@
                        clock-names = "cpu";
                        operating-points-v2 = <&cpu0_opp_table>;
                        #cooling-cells = <2>;
+                       clock-frequency = <1296000000>;
                };

                cpu2: cpu@2 {
@@ -98,6 +100,7 @@
                        clock-names = "cpu";
                        operating-points-v2 = <&cpu0_opp_table>;
                        #cooling-cells = <2>;
+                       clock-frequency = <1296000000>;
                };

                cpu3: cpu@3 {
@@ -108,6 +111,7 @@
                        clock-names = "cpu";
                        operating-points-v2 = <&cpu0_opp_table>;
                        #cooling-cells = <2>;
+                       clock-frequency = <1296000000>;
                };
        };
 
```

jayanta525 commented:
@sbilly 

- added, rebased and forced-pushed.

1715173329 commented:
Anything update?

hauke commented:
Why are the kernel and the u-boot device tree file different? We should try to use the same file for both.

jayanta525 commented:
Hello @hauke, thanks for the review. This PR is outdated and needs to be updated and rebased. I will do that ASAP.

jayanta525 commented:
@sbilly , can you help me out by testing some builds with new uboot and kernel?

hauke commented:
Could you also rename the patches, because these number are now already used by the board added in #3297 .

adschm commented on target/linux/sunxi/image/cortexa7.mk:
```diff
@@ -47,6 +47,16 @@ define Device/friendlyarm_nanopi-neo-air
+  DEVICE_VENDOR := FriendlyARM
+  DEVICE_MODEL := NanoPi R1
+  DEVICE_PACKAGES := kmod-rtc-sunxi kmod-usb2 kmod-usb-net-rtl8152 \
+	kmod-brcmfmac kmod-leds-gpio kmod-ledtrig-heartbeat wpad-basic \
```
wpad-basic-wolfssl should be used instead of wpad-basic now.

sbilly commented:
> @sbilly , can you help me out by testing some builds with new uboot and kernel?

sure. which commit?

jayanta525 committed:
sunxi: add support for FriendlyARM NanoPi R1

Specification:

- CPU: Allwinner H3, Quad-core Cortex-A7 Up to 1.2GHz
- DDR3 RAM: 512MB/1GB
- Network:
    10/100/1000M Ethernet x 1，
    10/100M Ethernet x 1
- WiFi: 802.11b/g/n, with SMA antenna interface
- USB Host: Type-A x2
- MicroSD Slot x 1
- MicroUSB: for OTG and power input
- Debug Serial Port: 3Pin 2.54mm pitch pin-header
- LED:
    nanopi:red:status
    nanopi:green:wan
    nanopi:green:lan
- KEY:
    reset
- Power Supply: DC 5V/2A

Installation:

- Write the image to SD Card with dd
- Boot NanoPi from the SD Card

Signed-off-by: Jayantajit Gogoi <jayanta.gogoi525@gmail.com>

jayanta525 commented:
@sbilly the latest forced-pushed commit.

sbilly commented:
> @sbilly the latest forced-pushed commit.

I'll test it in few days.

hauke commented:
I would prefer if you keep the kernel and the u-boot device tree files more or less the same.

Are you planning to add support for this board into upstream U-Boot and Linux kernel?

jayanta525 commented:
@hauke No, this board is pretty old 2019 Q1, and it's successor NanoPi R2S is already sent for upstream.

jayanta525 commented:
@hauke Updated, u-boot and kernel now uses the same device tree.

sbilly commented:
https://github.com/jayanta525/openwrt/commit/ac64bea7fb18c486190707989f39e9551dcfa07c  It's seem good to me. @jayanta525 

boot log: https://paste.ubuntu.com/p/jYpnFs23J6/

hauke commented:
Thank you, I applied the patches to master.

I removed the 300-sunxi-h3-suppress-clock-frequency-warning.patch patch, I am not sure if this is correct. Could you please send this upstream so it gets integrated into the upstream Linux kernel.

It would also be nice if you could send these patches to upstream Linux and U-Boot.
